sessiontxn/readcommitted/provider.go

```diff
@@ -0,0 +1,253 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package readcommitted
+
+import (
+	"context"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessiontxn"
+	"github.com/pingcap/tidb/table/temptable"
+	"github.com/tikv/client-go/v2/oracle"
+)
+
+type rcStmtContext struct {
+	sctx     sessionctx.Context
+	ctx      context.Context
+	prevStmt *rcStmtContext
```

Comment:
Would there be a race issue if some workers are still referencing fields in it and they are not finished, and the current statement is using it too.

Reply:
Can you give an example? In some cases, the executor will use the builder even the statement ends. But we fixed it by making the executor thread safe first: https://github.com/pingcap/tidb/pull/31956